mantidproject/mantid
# Fix crashes with WorkspaceGroup input to algorithms

Several algorithms turned out to have the same bug in `validateInputs()` that `AsymmetryCalc` was found to have in #15402. The input workspace was assumed to be a `MatrixWorkspace` or similar, and so the algorithm crashed when given a `WorkspaceGroup` as input. 

This PR fixes all the cases I could find, and adds tests that expose the bug and verify the fix:
- `AsymmetryCalc` (fixed in #15404 but this adds a test)
- `CalMuonDetectorPhases`
- `ConvertToDistribution`
- `ChangeTimeZero`
- `FFT`
- `MaxEnt`

**To test:**

_Code review:_
1. Try to find cases of this crash that I've missed - an algorithm with a `validateInputs()` method that uses the input workspace in its validation, and fails to check for null.
2. In some of the cases, would it be better for the algorithm to explicitly reject WorkspaceGroups rather than running once on each workspace?

_Functional tests:_
The new unit tests are intended to test the fixes here.
 To test manually, create a WorkspaceGroup in Mantid, highlight it in the MantidPlot workspace selector and then try to run the algorithm via the GUI. Mantid should not crash. (See description of #15402 for an example)

Fixes #15575 
Fixes #15433 

[Release notes](https://github.com/mantidproject/mantid/blob/15575_workspace_group_crashes/docs/source/release/v3.7.0/framework.rst) have been updated.



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
